src/NServiceBus.Transport.Msmq/MsmqMessageDispatcher.cs

```diff
@@ -77,6 +79,8 @@ void ExecuteTransportOperation(TransportTransaction transaction, UnicastTranspor
 
                         var label = GetLabel(message);
 
+                        await Task.Yield();
```

Comment:
You make a good point about the synchronization context. If you're testing on a console app, it won't make any difference, but if run from something that does have a synchronization context, that would mean you're resuming back on the calling thread...

Try testing it on a WinForms app or something and see how it looks. I think you might need to change to Task,Run after all.